src/lib/Lib/Filters.hpp

```diff
@@ -0,0 +1,109 @@
+//
+// This is a derivative work. originally part of the LLVM Project.
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// Copyright (c) 2023 Krystian Stasiowski (sdkrystian@gmail.com)
+//
+// Official repository: https://github.com/cppalliance/mrdox
+//
+
+#ifndef MRDOX_LIB_FILTERS_HPP
+#define MRDOX_LIB_FILTERS_HPP
+
+#include <mrdox/Platform.hpp>
+#include <span>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace clang {
+namespace mrdox {
+
+class FilterPattern
+{
+    // pattern without any wildcards
+    std::string raw_;
+    // pattern part lengths, where zero represents a wildcard
+    std::vector<std::size_t> parts_;
+
+#ifndef NDEBUG
+    std::string pattern_ = "*";
+#endif
+
+    /** Returns whether a string matches the multi-component pattern.
+
+        This function implements matching for patterns which contain
+        multiple components (i.e. both literal strings and wildcards).
+    */
+    bool
+    matchesSlow(
```

Comment:
I named it `matchSlow` because it implements a recursive matching algorithm for complex patterns i.e. it is the slow path. `match` itself contains the fast path & calls the slow path algorithm.

Reply:
I see. That's more problematic then because `matchSlow` is just a subroutine of `match`. 

So not only it describes how it does things rather than what it does (something like `matchParts`) but also there's no `matchFast` counterpart to it (so the description is opinionated) and assuming `match` is the fast variant of some common procedure would be wrong. 

The "fast path" is not really an alternative to `matchSlow` because they solve different problems. There's no intention to "match slowly" for some reason. 

They're more like `matchParts` and `matchLiteral`, implemented as "fast" as possible (even if it happens to be slow on a more difficult problem), which are subroutines of the more general `match` algorithm.

Reply:
If the grammar only supports `*`, it's worth noticing that string matching doesn't really require recursion. You just have to find the `*` to the left and right (they might be the same `*`), extract these as a literal prefix and suffix, "peel" these literals from a pattern string view and the input string from both sides, and repeat until the pattern is empty.

Reply:
The peeling algorithm will still end up having the same problems for inputs like `a*c*e`. The existing algorithm could be made iterative, but in general I find recursive algorithms to be easier for me to understand

Reply:
> same problems

What problems?

> easier for me to understand

I wasn't thinking in terms of design at all. What I mean is this "peeling algorithm" is not slow. If the recursive algorithm is equivalent to that, so it's also not slow. 

It was just one extra reason why the `matchSlow` name is inappropriate.